### ISSUE TITLE
IntelliSense client fails when using clang-x64 mode on Windows

### ISSUE THREAD

**Issue by Mossop** [microsoft/vscode-cpptools#3898]
**Type: LanguageService**

**Describe the bug**
- OS and Version: Windows 10 1903
- VS Code Version: 1.36.0
- C/C++ Extension Version: 0.24.0
- Other extensions you installed (and if the issue persists after disabling them): Tested with nothing but the extension I am developing.
- A clear and concise description of what the bug is.

I have been working on a configuration provider to pick up includes and defines from our custom build system (https://github.com/FractalBrew/vscode-mozillacpp). It is working well on Linux and OSX however on Windows it seems to cause the intellisense client to fail, but only when it tells it to use clang-x64 mode (which is unfortunate because that is the mode needed).

**To Reproduce**

Not straightforward unfortunately...

1. Build Firefox: https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Build_Instructions
2. Build and install the extension at https://github.com/FractalBrew/vscode-mozillacpp
3. Open the firefox source in VScode.
4. Make sure that "Mozilla" is chosen as the configuration provider.
5. Open the file docshell/base/nsDocShell.cpp

**Expected behavior**
Intellisense should work and show any diagnostics issues. Instead the logs indicate that intellisense is unavailable and it is falling back to the tag results.

**Additional context**

This is a log from trying this: https://gist.github.com/Mossop/af1bc103617eb62e95401ba16c1b9c15 (I stripped the tag parse updates).

As you can see the first problem is the message "Failed to read response from server: 22" and then it complains that the intellisense client is not available. Future use complains about the intellisense client being disconnected etc.

I tried attaching to the CPP process from visual studio and it didn't see any crashes.

**Comment by sean-mcmanus** [comment]
Can you attach to the Microsoft.VSCode.CPP.IntelliSense.Msvc.exe process after opening the file? If it crashes too fast, you can cut all the contents, attach, and then paste the contents to make it crash.

We could/should also investigate the repro.

It’s with clang-cl.exe – we don’t support that scenario very well currently. You could try adding -fms-extensions to the end of the compilerPath.

**Comment by Mossop** [comment]
I don't currently pass the compiler path in the configuration. It seems to confuse intellisense into thinking it is dealing with a WSL compiler and switches to gcc intellisense mode.

I managed to catch the crash:

```
Exception thrown: read access violation.
**_locator_for_curr_id** was nullptr

>	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!scan_expr_full()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!scan_expr_into_new_init_component()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!scan_expr_list()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!scan_call_arguments()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!scan_function_call()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!scan_expr_full()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!scan_expr_into_new_init_component()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!scan_expr_as_init_component()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!scan_init_component_with_potential_pack_expansion()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!_scan_full_initializer_expr_as_component()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!expr_init_scalar_variable()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!_initializer()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!variable_declaration()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!_scan_nonmember_declaration()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!_declaration()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!_translation_unit()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!_process_translation_unit()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!_cfe_main()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!_cfe_main_exception_handler()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!_edg_main()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!invoke_edge_compiler()	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!edge_compiler_main(int,char const * * const)	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!preparse(int,char const * *,class a_scout_store *,class edge::translation_unit *)	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!a_compiler_thread::compiler_thread_routine(class a_compiler_thread *)	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!std::_LaunchPad<class std::unique_ptr<class std::tuple<void (*)(class a_compiler_thread *),class a_compiler_thread *>,struct std::default_delete<class std::tuple<void (*)(class a_compiler_thread *),class a_compiler_thread *> > > >::_Run(class std::_LaunchPad<class std::unique_ptr<class std::tuple<void (*)(class a_compiler_thread *),class a_compiler_thread *>,struct std::default_delete<class std::tuple<void (*)(class a_compiler_thread *),class a_compiler_thread *> > > > *)	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!std::_LaunchPad<class std::unique_ptr<class std::tuple<void (*)(class a_compiler_thread *),class a_compiler_thread *>,struct std::default_delete<class std::tuple<void (*)(class a_compiler_thread *),class a_compiler_thread *> > > >::_Go(void)	Unknown
 	Microsoft.VSCode.CPP.IntelliSense.Msvc.exe!thread_start<unsigned int (__stdcall*)(void *)>()	Unknown
 	kernel32.dll!@BaseThreadInitThunk@12()	Unknown
 	ntdll.dll!__RtlUserThreadStart()	Unknown
 	ntdll.dll!__RtlUserThreadStart@8()	Unknown
```

I'm going to see if I can narrow this down some to make this easier to reproduce.

**Comment by sean-mcmanus** [comment]
From your logs, it's using clang-cl.exe with clang-x64 mode, which has some known bugs such as https://github.com/microsoft/vscode-cpptools/issues/3296 if -fms-extensions isn't added. You could also try msvc-x64 mode (it could give better results depending on what you're compiling).

Yeah, for a crash like that, our parser team usually needs an "isolated repro" in order to fix (i.e. with no `#includes`). Since the repro project seems to be available, we might be able to obtain an isolated repro if you are not able to (it can get tricky if libraries are required).

**Comment by Mossop** [comment]
It's not clear to me where I would add -fms-extensions when I am not passing the compiler path.

Unfortunately the code uses some clang internal functions that aren't recognised by msvc-x64 mode.

**Comment by sean-mcmanus** [comment]
From the logs you provided, it doesn't appear that the custom configuration provider is being used and instead it shows it's using 
Attempting to get defaults from compiler in compile_commands.json file: 'c:/Users/Dave/.mozbuild/clang/bin/clang-cl.exe' . So if you set the compilerPath to that, then you'd add the -fms-extensions afterwards. You could also set the compilerPath via the config provider API (SourceFileConfiguration.compilerPath).

**Comment by Mossop** [comment]
As I said, when I provide it by SourceFileConfiguration.compilerPath it decides that it is a WSL compiler and adds a bunch of unwanted includes to the list.

**Comment by sean-mcmanus** [comment]
Oh, you found a bug with how we handle clang-cl.exe when it's set as the compilerPath -- we incorrectly do not query the compiler for system includes/defines like clang.exe and instead treat it like cl.exe and because the intelliSenseMode is not msvc-x64 we "error out" and try to find another valid clang-x64 compatible compiler, which is the WSL compiler. I've filed a new bug (https://github.com/microsoft/vscode-cpptools/issues/3906), since it's unrelated to this issue. A potential workaround would be to use clang.exe as the compilerPath and pass -fms-extensions (and possibly a bunch of other flags that clang-cl.exe internally forwards to clang.exe) instead of using clang-cl.exe directly.

**Comment by Colengms** [comment]
@Mossop , Could you provide your compile_commands.json ?

**Comment by Mossop** [comment]
Sorry I deleted it earlier. This issue reproduces with no compile_commands.json. I should have some time tomorrow to try to narrow the issue down tomorrow but it may take some time since the file in question is quite large.

**Comment by sean-mcmanus** [comment]
In your logs at https://gist.github.com/Mossop/af1bc103617eb62e95401ba16c1b9c15 , do you know how the clang system defines got obtained? Because we're seeing clang-cl.exe give errors when we pass it clang-specific flags used to query for the system defines.

in other words, I'm seeing the call to
`clang-cl.exe -Wp,-v -E -xc -dD -x c++ nul`
fail, but it works with clang.exe.

My guess would be that the compile_commands.json has some special flags that are being passed to clang-cl.exe that somehow causes it to work.

**Comment by Mossop** [comment]
I've narrowed this down a bit to a simple c++ file that causes it: https://gist.github.com/Mossop/c80a38549b03e732f1410bbc917c4e1d

It's a bit tricky to narrow this down further though. Removing any part of that C++ file makes the crash go away. The file it is including is empty. Removing some of the directories in the includePath will fix it, removing others won't but there seems to be some combination of those includes that causes the crash so a bisect won't work. I've tried to set this up in a standalone directory without my extension involved and it doesn't reproduce there so either I'm missing some detail about what my extension is doing or this only happens with a custom configuration. I could try writing a far simpler custom configuration extension if that is useful but that would take me longer.

I've been using `clang-cl.exe -std:c++14 -TP -v -E -Xclang -dM nul` to get the compiler defaults.

**Comment by Mossop** [comment]
FWIW I have since switched to using msvc intellisense mode and got things working somewhat better so I am less blocked by this bug, though clang mode seems to work for cases where msvc mode does not so it would still be useful to see this fixed.

https://github.com/FractalBrew/vscode-mozillacpp/commit/f286dd0b44ac95b71b658e39d87c58763e4b0c0c is the version of my extension that this issue reproduces with.

**Comment by Colengms** [comment]
@Mossop , I'm able to repro the same exception stack you posted.  I'd like to pare this down to a simple repro, to file a bug against the IntelliSense folks.  It looks like your environment does not match what I've set up following the Mozilla build steps you linked.  I'm not able to resolve "mozilla/dom/ContentChild.h", for example.  (I see it under "mozilla-central\dom\ipc\ContentChild.h" and "mozilla-central\obj-x86_64-pc-mingw32\dist\include\mozilla\dom\ContentChild.h", but it's unclear if either of those are correct)

Could you generate preprocessed output for the repro'ing source file?  Passing -E to clang should do that.  And, could you confirm the issue still repros with the resulting output, and send that?  That would remove header resolution from the equation, and give us one big file to chop large sections out of to widdle this down to the offending few lines in.  :)

**Comment by Mossop** [comment]
Ok, I've managed to generate a folder that reproduces the crash, no extensions other than cpptools needed: https://github.com/Mossop/test_cpptools. It had a hardcoded path for my clang-cl so that needs changing.

Hopefully that helps.

**Comment by Colengms** [comment]
Thanks @Mossop !  That is exactly what I needed, and saved me a lot of time trying to slim it down.  I was able to use that to repro the issue in VS as well.  I've opened a bug against VS.  (We share IntelliSense code with VS).  That is here, in case you want to up-vote it: https://developercommunity.visualstudio.com/content/problem/643597/vcpgksrv-crash-in-clang-mode.html

Note that although the top of the stack is the same, the stack from the last code you posted is actually somewhat different from the original issue.  It's possible there may be multiple issues here.
